File: dfs/longest_non_repeating_path/longest_non_repeating_path.cpp
#include <vector>
#include <unordered_map>
#include <list>
#include <numeric>
#include <string>
using namespace std;

class LongestNonRepeatingPath
{ // LeetCode Q.2246.
private:
    string chars;
    unordered_map<int, list<int>> tree;
    int maxNonRepeatingLen;

    int dfsMaxNonRepeatingLen(int idx)
    {
        if (tree.find(idx) == tree.end())
        { // Base case: leave nodes.
            return 1;
        }

        list<int> top_2_children_lens;
        for (auto child_idx : tree[idx])
        {
            top_2_children_lens.push_back(0); // Default to 0.
            int children_len = dfsMaxNonRepeatingLen(child_idx);
            if (chars[child_idx] != chars[idx])
            {
                top_2_children_lens.back() += children_len;
            }

            if (top_2_children_lens.size() > 2)
            { // Always keep top 2.
                top_2_children_lens.sort();
                top_2_children_lens.pop_front();
            }
        }

        top_2_children_lens.sort();
        int straight_len = 1 + top_2_children_lens.back();
        // Curve path is centered at current node.
        int curve_len = accumulate(top_2_children_lens.begin(), top_2_children_lens.end(), 1);

        if (max(straight_len, curve_len) > maxNonRepeatingLen)
        {
            maxNonRepeatingLen = max(straight_len, curve_len);
        }
        return straight_len; // Only straight len can be reported to parent.
    }

public:
    int findLongestNonRepeatingLen(vector<int> &parent, string characters)
    {
        chars = characters;
        tree.clear();
        for (int childIdx = 0; childIdx < parent.size(); childIdx++)
        {
            int parentIdx = parent[childIdx];
            if (parentIdx != -1)
            { // Non root nodes.
                if (tree.find(parentIdx) == tree.end())
                {
                    tree[parentIdx] = {};
                }
                tree[parentIdx].push_back(childIdx);
            }
        }

        maxNonRepeatingLen = 1;   // Base case.
        dfsMaxNonRepeatingLen(0); // Start from root.
        return maxNonRepeatingLen;
    }
};

File: dfs/longest_non_repeating_path/longest_non_repeating_path.py

class LongestNonRepeatingPath:  # LeetCode Q.2246.
    def __init__(self):
        self.string = ""
        self.tree: dict[int, list[int]] = dict()
        self.max_non_repeating_len = 1  # Base case.

    def find_longest_non_repeating_len(self, parent: list[int], string: str) -> int:
        self.string = string
        self.tree.clear()
        for child_idx, parent_idx in enumerate(parent):
            if parent_idx != -1:  # Non root nodes.
                if parent_idx not in self.tree.keys():
                    self.tree.update({parent_idx: []})
                self.tree[parent_idx].append(child_idx)

        self.max_non_repeating_len = 1  # Base case.
        self._dfs_max_non_repeating_len(0)  # Start from root.
        return self.max_non_repeating_len

    def _dfs_max_non_repeating_len(self, idx: int) -> int:
        if idx not in self.tree.keys():  # Base case: leave nodes.
            return 1

        top_2_children_lens = []
        for child_idx in self.tree[idx]:
            top_2_children_lens.append(0)  # Default to 0.

            children_len = self._dfs_max_non_repeating_len(child_idx)
            if self.string[child_idx] != self.string[idx]:
                top_2_children_lens[-1] += children_len

            if len(top_2_children_lens) > 2:  # Always keep top 2.
                top_2_children_lens.sort(reverse=True)
                top_2_children_lens.pop(-1)

        straight_len = 1 + max(top_2_children_lens)
        # Curve path is centered at current node.
        curve_len = 1 + sum(top_2_children_lens)

        if max(straight_len, curve_len) > self.max_non_repeating_len:
            self.max_non_repeating_len = max(straight_len, curve_len)

        return straight_len  # Only straight len can be reported to parent.
